modules/photo/src/contrast_preserve.hpp

```diff
@@ -81,350 +83,331 @@ int round_num(double a)
                     vector <double> curGrad;
-                    gradvector(curIm,curGrad);
-                    add_to_vector_poly(polyGrad,curGrad,idx1);
+                    gradvector(curIm, curGrad);
+                    add_to_vector_poly(polyGrad, curGrad, idx1);
                 }
             }
 }
 
-void Decolor::wei_update_matrix(vector < vector <double> > &poly, vector <double> &Cg, Mat &X)
+void Decolor::wei_update_matrix(const vector<vector<double> > &poly, vector<double> &Cg, Mat &X)
 {
-    int size = static_cast<int>(poly.size()), size0 = static_cast<int>(poly[0].size());
+    const int size = static_cast<int>(poly.size());
+    const int size0 = static_cast<int>(poly[0].size());
     Mat P = Mat(size, size0, CV_32FC1);
-    Mat A = Mat(size, size, CV_32FC1);
 
     for (int i = 0; i < size; i++)
-        for (int j = 0; j < size0;j++)
-            P.at<float>(i,j) = (float) poly[i][j];
+        for (int j = 0; j < size0; j++)
+            P.at<float>(i, j) = static_cast<float>(poly[i][j]);
 
-    Mat P_trans = P.t();
+    const Mat P_trans = P.t();
     Mat B = Mat(size, size0, CV_32FC1);
-    for(int i =0;i < size;i++)
+    for (int i = 0; i < size; i++)
     {
-        for(int j = 0, end = (int)Cg.size(); j < end;j++)
-            B.at<float>(i,j) = (float) (poly[i][j] * Cg[j]);
+        for (int j = 0, end = static_cast<int>(Cg.size()); j < end; j++)
+            B.at<float>(i, j) = static_cast<float>(poly[i][j] * Cg[j]);
     }
 
-    A = P*P_trans;
+    const Mat A = P * P_trans;
     solve(A, B, X, DECOMP_NORMAL);
-
 }
 
-void Decolor::wei_inti(vector < vector <int> > &comb, vector <double> &wei)
+void Decolor::wei_inti(const vector<Vec3i> &comb, vector<double> &wei)
 {
-    vector <double> initRGB;
-
-    initRGB.push_back( .33 );
-    initRGB.push_back( .33 );
-    initRGB.push_back( .33 );
-    wei = product(comb,initRGB);
+    double initRGB[3] = { .33, .33, .33 };
+    wei = product(comb, initRGB);
 
-    vector <int> sum;
+    vector<int> sum(comb.size());
 
-    for(unsigned int i=0;i<comb.size();i++)
-        sum.push_back(comb[i][0] + comb[i][1] + comb[i][2]);
+    for (unsigned int i = 0; i<comb.size(); i++)
+        sum[i] = (comb[i][0] + comb[i][1] + comb[i][2]);
 
-    for(unsigned int i=0;i<sum.size();i++)
+    for (unsigned int i = 0; i<sum.size(); i++)
     {
-        if(sum[i] == 1)
+        if (sum[i] == 1)
             wei[i] = wei[i] * double(1);
         else
             wei[i] = wei[i] * double(0);
     }
-
-    initRGB.clear();
-    sum.clear();
-
 }
 
-void Decolor::grayImContruct(vector <double> &wei, Mat img, Mat &Gray)
+void Decolor::grayImContruct(const vector<double> &wei, const Mat &img, Mat &Gray) const
 {
-    int h=img.size().height;
-    int w=img.size().width;
+    const int h = img.rows;
+    const int w = img.cols;
 
-    vector <Mat> rgb_channel;
-    split(img,rgb_channel);
+    Mat rgb_channel[3];
+    split(img, rgb_channel);
```

Comment:
> Mat rgb_channel[3];

This change is dangerous. There is no checks number of channels in this function.
RGBA/BGRA image will corrupt memory.

Also this change has minimal effect on performance.

The same note for other similar changes.